### PR TITLE
feat(nexus/resize): implement ResourceResize for nexus

### PR DESCRIPTION
**Pull request by dsharma-dc**

### CODE REVIEW

**Comment by dsharma-dc** [comment]
bors try

**Comment by dsharma-dc** [comment]
> ## try
> Build failed:
> 
> * [continuous-integration/jenkins/branch](https://mayastor-ci.mayadata.io/job/Mayastor-Control-Plane/job/trying/39/display/redirect)

Dataplane image not available yet with changes merged from today it seems. gRPC fails with `Unimplemented`

File: control-plane/agents/src/bin/core/volume/specs.rs [hunk 1]
```diff
@@ -320,18 +320,6 @@ pub(crate) async fn healthy_volume_replicas(
     }
 }
 
-/// Check if the volume state is ok for resize operation.
-pub fn vol_status_ok_for_resize(spec: &VolumeSpec) -> Result<(), SvcError> {
```
Comment (dsharma-dc) [inline control-plane/agents/src/bin/core/volume/specs.rs:324]: Removed this function altogether as we are allowing expansion of published and unpublished, both types of volume.

**Comment by dsharma-dc** [comment]
bors try

**Comment by dsharma-dc** [comment]
bors cancel

File: tests/bdd/features/volume/resize/test_resize_online.py
Comment (dsharma-dc) [inline tests/bdd/features/volume/resize/test_resize_online.py:1]: need to remove prints from the file

File: control-plane/agents/src/bin/core/controller/io_engine/v1/nexus.rs [hunk 2]
```diff
@@ -118,6 +118,46 @@ impl crate::controller::io_engine::NexusApi<()> for super::RpcClient {
...
+            });
+
+        // We map certain tonic gRPC error codes to NexusResizeStatusUnknown, assuming that
+        // we don't know where the nexus resize request went wrong.
+        let rpc_nexus = match result {
+            Err(status)
+                if status.tonic_code() == tonic::Code::Unknown
+                    || status.tonic_code() == tonic::Code::DeadlineExceeded
+                    || status.tonic_code() == tonic::Code::Internal
+                    || status.tonic_code() == tonic::Code::Unavailable =>
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/io_engine/v1/nexus.rs]: nit: perhaps neater

```suggestion
                if matches!(status.tonic_code(), Code::Unknown | Code::DeadlineExceeded | Code::Internal | Code::Unavailable) =>
```

File: control-plane/agents/src/bin/core/controller/io_engine/v1/nexus.rs [hunk 3]
```diff
@@ -118,6 +118,46 @@ impl crate::controller::io_engine::NexusApi<()> for super::RpcClient {
...
+            {
+                tracing::warn!(nexus=%request.uuid, status=%status, "Resize status unclear.");
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/io_engine/v1/nexus.rs]:
```suggestion
                tracing::warn!(nexus.uuid=%request.uuid, %status, "Resize status unclear");
```

File: control-plane/agents/src/bin/core/controller/io_engine/v1/nexus.rs [hunk 4]
```diff
@@ -118,6 +118,46 @@ impl crate::controller::io_engine::NexusApi<()> for super::RpcClient {
...
+            _else => _else,
+        }?;
+
+        match rpc_nexus.into_inner().nexus {
+            Some(nexus) => Ok(rpc_nexus_to_agent(&nexus, &request.node)?),
+            None => Err(SvcError::Internal {
+                details: format!(
+                    "resource: {}, request: {}, err: {}",
+                    "Nexus", "resize_nexus", "no nexus returned"
+                ),
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/io_engine/v1/nexus.rs]:
```suggestion
                details: "resource: Nexus, request: resize_nexus, error: no nexus returned".into(),
```

File: control-plane/agents/src/bin/core/controller/reconciler/nexus/mod.rs [hunk 5]
```diff
@@ -606,3 +606,37 @@ pub(super) async fn faulted_nexus_remover(
...
+    nexus: &mut OperationGuardArc<NexusSpec>,
+    volume: &OperationGuardArc<VolumeSpec>,
+    context: &PollContext,
+) -> PollResult {
+    let registry = context.registry();
+    let nexus_state = registry.nexus(nexus.as_ref().uid()).await?;
+    let required_size = volume.as_ref().size;
+
+    // Nexus's size doesn't need any fixup.
+    if nexus.as_ref().size == required_size {
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/nexus/mod.rs:623]: This should be done before getting the nexus state

File: control-plane/agents/src/bin/core/controller/reconciler/nexus/mod.rs [hunk 6]
```diff
@@ -606,3 +606,37 @@ pub(super) async fn faulted_nexus_remover(
...
+        debug_assert_eq!(nexus_state.size, nexus.as_ref().size);
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/nexus/mod.rs]: Don't think this is valid, this scenario can always happen if a grpc times out and we don't know if it reached the io-engine or not
```suggestion
```

File: control-plane/agents/src/bin/core/tests/volume/resize.rs
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/tests/volume/resize.rs:1]: Do we need a test to validate the resize reconciler?

File: control-plane/agents/src/bin/core/volume/operations_helper.rs [hunk 7]
```diff
@@ -554,6 +554,23 @@ impl OperationGuardArc<VolumeSpec> {
...
     }
 
+    /// Resize the volume target(nexus).
+    pub(super) async fn resize_volume_target(
+        &self,
+        registry: &Registry,
+        nexus_grd: &mut OperationGuard<ResourceMutex<NexusSpec>, NexusSpec>,
+        requested_size: u64,
+    ) -> Result<(), SvcError> {
+        let nexus = registry.nexus(nexus_grd.as_ref().uid()).await?;
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations_helper.rs]: Why is this needed? nexux_grd already has node and uuid?

File: control-plane/agents/src/bin/core/volume/operations_helper.rs [hunk 8]
```diff
@@ -554,6 +554,23 @@ impl OperationGuardArc<VolumeSpec> {
         Ok(())
     }
 
+    /// Resize the volume target(nexus).
+    pub(super) async fn resize_volume_target(
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations_helper.rs]:
```suggestion
    pub(super) async fn resize_target(
```

File: control-plane/agents/src/bin/core/volume/operations_helper.rs [hunk 9]
```diff
@@ -554,6 +554,23 @@ impl OperationGuardArc<VolumeSpec> {
...
+        &self,
+        registry: &Registry,
+        nexus_grd: &mut OperationGuard<ResourceMutex<NexusSpec>, NexusSpec>,
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/operations_helper.rs]:
```suggestion
        nexus_grd: &mut OperationGuardArc<NexusSpec>,
```

File: control-plane/agents/src/common/errors.rs [hunk 10]
```diff
@@ -96,6 +96,15 @@ pub enum SvcError {
     InvalidPoolDeviceNum { disks: Vec<PoolDeviceUri> },
     #[snafu(display("Nexus '{}' not found", nexus_id))]
     NexusNotFound { nexus_id: String },
+    #[snafu(display(
+        "Nexus '{}' resize status not determined. Requested size {}",
+        nexus_id,
+        requested_size
+    ))]
```
Comment (tiagolobocastro) [inline control-plane/agents/src/common/errors.rs]:
```suggestion
        "Nexus '{nexus_id}' resize status not determined. Requested size {requested_size}",
))]
```

File: control-plane/stor-port/src/types/v0/transport/nexus.rs [hunk 11]
```diff
@@ -242,6 +242,29 @@ pub struct CreateNexus {
     pub config: Option<NexusNvmfConfig>,
 }
 
+/// Resize Nexus Request
```
Comment (tiagolobocastro) [inline control-plane/stor-port/src/types/v0/transport/nexus.rs]:
```suggestion
/// A request to resize a Nexus.
```

File: control-plane/stor-port/src/types/v0/transport/nexus.rs [hunk 12]
```diff
@@ -242,6 +242,29 @@ pub struct CreateNexus {
...
+#[derive(Serialize, Deserialize, Default, Debug, Clone, Eq, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct ResizeNexus {
+    /// id of the io-engine instance
```
Comment (tiagolobocastro) [inline control-plane/stor-port/src/types/v0/transport/nexus.rs]:
```suggestion
    /// Id of the io-engine instance.
```

File: control-plane/stor-port/src/types/v0/transport/nexus.rs [hunk 13]
```diff
@@ -242,6 +242,29 @@ pub struct CreateNexus {
...
+    pub node: NodeId,
+    /// uuid of the nexus
```
Comment (tiagolobocastro) [inline control-plane/stor-port/src/types/v0/transport/nexus.rs]:
```suggestion
    /// Uuid of the nexus to be resized.
```

File: control-plane/stor-port/src/types/v0/transport/nexus.rs [hunk 14]
```diff
@@ -242,6 +242,29 @@ pub struct CreateNexus {
...
+    pub uuid: NexusId,
+    /// requested size of nexus
```
Comment (tiagolobocastro) [inline control-plane/stor-port/src/types/v0/transport/nexus.rs]:
```suggestion
    /// Requested size for the nexus.
```

File: tests/bdd/features/volume/resize/resize_online.feature [hunk 15]
```diff
@@ -0,0 +1,95 @@
+Feature: Volume resize
+
+    Background:
+        Given a deployer cluster
+
+    Scenario Outline: Expand a volume
+        Given a <publish_status> volume with <repl_count> replicas and all are healthy
```
Comment (tiagolobocastro) [inline tests/bdd/features/volume/resize/resize_online.feature]: unpublished is already covered by the offline resize?

```suggestion
        Given a published volume with <repl_count> replicas and all are healthy
```

File: control-plane/agents/src/bin/core/tests/volume/resize.rs
Comment (dsharma-dc) [reply control-plane/agents/src/bin/core/tests/volume/resize.rs:1]: > Do we need a test to validate the resize reconciler?

I had tried testing it by introducing artificial drop via blocking grpc to nexus io-engine. However, it doesn't produce expected behaviour of uncertainty(no-response) because packet is dropped with certainty and request never reaches io-engine. 
I'm adding resize operation to `ReplicaOperations` and `NexusOperations` which will help us put the volume in a state to be picked up by resize reconcilers.

File: control-plane/agents/src/bin/core/controller/reconciler/nexus/mod.rs (hunk 5 above)
Comment (dsharma-dc) [reply control-plane/agents/src/bin/core/controller/reconciler/nexus/mod.rs:623]: > This should be done before getting the nexus state

reordered

File: control-plane/agents/src/bin/core/controller/io_engine/v1/nexus.rs (hunk 2 above)
Comment (dsharma-dc) [reply control-plane/agents/src/bin/core/controller/io_engine/v1/nexus.rs]: > nit: perhaps neater

modified

File: control-plane/agents/src/bin/core/volume/operations_helper.rs (hunk 7 above)
Comment (dsharma-dc) [reply control-plane/agents/src/bin/core/volume/operations_helper.rs]: Yeah, I think earlier this function was doing something more that needed state, but is now simplified and this remained. Modified to use from nexus_grd now but need to clone, otherwise it complains of immutable borrow attempt 
if directly passed in as `&nexus_grd.lock().node`
```
        let node = nexus_grd.lock().node.clone();
        let nexus_uuid = nexus_grd.uuid().clone();
        nexus_grd
            .resize(
                registry,
                &ResizeNexus::new(&node, &nexus_uuid, requested_size),
            )
            .await?;

```

**Comment by dsharma-dc** [comment]
bors try

**Comment by dsharma-dc** [comment]
@tiagolobocastro Comments are addressed, ptal if we are good to go.

File: control-plane/agents/src/bin/core/volume/operations_helper.rs (hunk 7 above)
Comment (tiagolobocastro) [reply control-plane/agents/src/bin/core/volume/operations_helper.rs]: Can use nexus_grd.as_ref().node.clone();

File: control-plane/agents/src/bin/core/controller/io_engine/v1/nexus.rs [hunk 16]
```diff
@@ -118,6 +118,46 @@ impl crate::controller::io_engine::NexusApi<()> for super::RpcClient {
...
+            Err(status)
+                if matches!(
+                    status.tonic_code(),
+                    tonic::Code::Unknown
+                        | tonic::Code::DeadlineExceeded
+                        | tonic::Code::Internal
+                        | tonic::Code::Unavailable
+                ) =>
+            {
+                tracing::warn!(nexus.uuid=%request.uuid, status=%status, "Resize status unclear.");
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/io_engine/v1/nexus.rs]:
```suggestion
                tracing::warn!(nexus.uuid=%request.uuid, status=%status, "Resize status unclear");
```

File: control-plane/stor-port/src/types/v0/transport/nexus.rs [hunk 17]
```diff
@@ -242,6 +242,29 @@ pub struct CreateNexus {
     pub config: Option<NexusNvmfConfig>,
 }
 
+/// A request to resize a Nexus.
+#[derive(Serialize, Deserialize, Default, Debug, Clone, Eq, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct ResizeNexus {
+    /// Id of the io-engine instance
```
Comment (tiagolobocastro) [inline control-plane/stor-port/src/types/v0/transport/nexus.rs]: Pls change others as well to end with `.`

```suggestion
    /// Id of the io-engine instance.
```

File: tests/bdd/features/volume/resize/resize_online.feature [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+        And the volume should be expanded to the new capacity
+        And the new capacity should be available for the application
+        Examples:
+        | repl_count |
+        | 1          |
+        | 3          |
+
+
+    Scenario Outline: Expand a volume with an offline replica
+        Given a <publish_status> volume with more than one replicas
```
Comment (tiagolobocastro) [inline tests/bdd/features/volume/resize/resize_online.feature]: Same thing here and further below as well, isn't unpublished already covered by offline?

**Review by tiagolobocastro — APPROVED** [review]

File: tests/bdd/features/volume/resize/resize_online.feature (hunk 18 above)
Comment (dsharma-dc) [reply tests/bdd/features/volume/resize/resize_online.feature]: Segregated.

File: control-plane/stor-port/src/types/v0/transport/nexus.rs (hunk 17 above)
Comment (dsharma-dc) [reply control-plane/stor-port/src/types/v0/transport/nexus.rs]: > Pls change others as well to end with `.`

Modified. However, rest of the structs don't have it in the comments.

Comment (tiagolobocastro) [reply control-plane/stor-port/src/types/v0/transport/nexus.rs]: They're probably "older", before we started following  this practice

**Review by Abhinandan-Purkait — APPROVED** [review]

**Comment by dsharma-dc** [comment]
bors merge
